kf/nfconfig.go

```diff
@@ -942,3 +943,311 @@ func getHostInterfaces() (map[string]bool, error) {
 	}
 	return hostIfaces, nil
 }
+
+func (c *NFConfigs) AddAndStartBPF(bpfProg *models.BPFProgram, ifaceName string, direction string) error {
+	var bpfList *list.List
+	if bpfProg == nil {
+		return fmt.Errorf("AddAndStartBPF - bpf program is nil")
+	}
+
+	if bpfProg.AdminStatus == models.Disabled {
+		return nil
+	}
+
+	switch direction {
+	case models.XDPIngressType:
+		bpfList = c.IngressXDPBpfs[ifaceName]
+	case models.IngressType:
+		bpfList = c.IngressTCBpfs[ifaceName]
+	case models.EgressType:
+		bpfList = c.EgressTCBpfs[ifaceName]
+	default:
+		return fmt.Errorf("unknown direction type")
+	}
+
+	for e := bpfList.Front(); e != nil; e = e.Next() {
+		data := e.Value.(*BPF)
+		if data.Program.Name == bpfProg.Name {
+			log.Warn().Msgf("%v is already running on %v iface and in %v direction ", data.Program.Name, ifaceName, direction)
+			return nil
+		}
+
+		if data.Program.SeqID == bpfProg.SeqID {
+			log.Warn().Msgf("duplicate seq Id detected for %v in direction %v", data.Program.Name, direction)
+			return nil
+		}
+	}
+	for e := bpfList.Front(); e != nil; e = e.Next() {
+		data := e.Value.(*BPF)
+		if data.Program.SeqID > bpfProg.SeqID {
+			bpf := NewBpfProgram(c.ctx, *bpfProg, c.HostConfig)
+			tmpBPF := bpfList.InsertBefore(bpf, e)
+			if err := c.DownloadAndStartBPFProgram(tmpBPF, ifaceName, direction); err != nil {
+				return fmt.Errorf("failed to download and start eBPF program %s version %s iface %s direction %s", bpfProg.Name, bpfProg.Version, ifaceName, direction)
```

Comment:
Is there a coding guideline to understand how many characters can a line go to?

Reply:
https://github.com/l3af-project/l3afd/blob/main/docs/CONTRIBUTING.md#coding-style says to "See [uber-go](https://github.com/uber-go/guide/blob/master/style.md)" which says [here](https://github.com/uber-go/guide/blob/master/style.md#avoid-overly-long-lines):

> Avoid lines of code that require readers to scroll horizontally or turn their heads too much.
> 
> We recommend a soft line length limit of 99 characters. Authors should aim to wrap lines before hitting this limit, but it is not a hard limit. Code is allowed to exceed this limit.